src/opera/parser/tosca/csar.py

```diff
@@ -97,109 +344,35 @@ def package_csar(self, output, service_template=None, csar_format="zip"):
-                        root_yaml_template = yaml.safe_load(root_template)
-                        self._metadata = root_yaml_template.get('metadata')
-
-                    self._get_template_version()
-                    self._get_template_name()
-
-                    return root_yaml_files[0].name
-            except Exception as e:
-                raise ParseError("Invalid CSAR structure: {}".format(e), self)
-
-    def _validate_csar_version(self):
-        csar_version = self._tosca_meta.get('CSAR-Version')
-        if csar_version and csar_version != 1.1:
-            raise ParseError('CSAR-Version entry in the CSAR {} is '
-                             'required to denote version 1.1".'.
-                             format(self._csar_name), self)
-        return csar_version
-
-    def _validate_tosca_meta_file_version(self):
-        tosca_meta_file_version = self._tosca_meta.get(
-            'TOSCA-Meta-File-Version')
-        if tosca_meta_file_version and tosca_meta_file_version != 1.1:
-            raise ParseError(
-                'TOSCA-Meta-File-Version entry in the CSAR {} is '
-                'required to denote version 1.1".'.format(
-                    self._csar_name), self)
-        return tosca_meta_file_version
-
-    def _get_entry_definitions(self):
-        if 'Entry-Definitions' not in self._tosca_meta:
-            raise ParseError(
-                'The CSAR "{}" is missing the required metadata '
-                '"Entry-Definitions" in "TOSCA-Metadata/TOSCA.meta".'.format(
-                    self._csar_name), self)
-        return self._tosca_meta.get('Entry-Definitions')
-
-    # prepared for future, currently not used since this is an optional keyname
-    def _get_created_by(self):
-        return self._tosca_meta.get('Created-By')
-
-    # prepared for future, currently not used since this is an optional keyname
-    def _get_other_definitions(self):
-        return self._tosca_meta.get('Other-Definitions')
-
-    def _get_template_version(self):
-        if "template_version" not in self._metadata:
-            raise Exception(
-                'The CSAR "{}" is missing the required '
-                'template_version in metadata".'.format(self._csar_name))
-        return self._metadata.get('template_version')
-
-    def _get_template_name(self):
-        if "template_version" not in self._metadata:
-            raise ParseError(
-                'The CSAR "{}" is missing the required '
-                'template_name in metadata".'.format(self._csar_name), self)
-        return self._metadata.get('template_name')
-
-    # prepared for future, currently not used since this is an optional keyname
-    def _get_author(self):
-        return self._metadata.get('template_author')
+        raise OperaError("Cannot unpackage a directory-based CSAR.")
+
+    def members(self) -> List[PurePath]:
+        # queue loop instead of recursion
+        # processed by preorder traversal
```

Comment:
I'm curious - why preorder traversal? Is this even important?

Reply:
It's so you know roughly what order the files are in in the result. Reading this comment is simpler than reading the recursion-but-actually-loop below. :)

Reply:
Oh, very smart move.